db/collab.go

```diff
@@ -53,19 +61,69 @@ func (s *Session) RemoveConn(uid string) error {
 		return errors.New("Could not remove unconnected user")
 	}
 	delete(s.Conns, uid)
+	// Stop other connections from sending to removed connection
+	for _, conn := range s.Conns {
+		delete(conn.Subscriptions, uid)
+	}
 	return nil
 }
 
-func (s *Session) Broadcast(msg Message) error {
-	var mostRecentError error
+func (s *Session) BroadcastAll(msg Message) (lastErr error) {
 	s.Lock.Lock()
 	defer s.Lock.Unlock()
 	for _, conn := range s.Conns {
-		if err := websocket.JSON.Send(conn, msg); err != nil {
-			mostRecentError = err
+		if err := websocket.JSON.Send(conn.Conn, msg); err != nil {
+			lastErr = err
+		}
+	}
+	return
+}
+
+func (s *Session) BroadcastError(uid string, err string) error {
+	errorMsg := Message{
+		Author: uid,
+		Type:   "ERROR",
+		Body:   err,
+	}
+	if broadcastErr := s.BroadcastTo(errorMsg, uid); broadcastErr != nil {
+		return broadcastErr
+	}
+	return nil
+}
+
+func (s *Session) BroadcastTo(msg Message, uids ...string) (lastErr error) {
+	s.Lock.Lock()
+	defer s.Lock.Unlock()
+	for _, uid := range uids {
+		if err := websocket.JSON.Send(s.Conns[uid].Conn, msg); err != nil {
+			lastErr = err
 		}
 	}
-	return mostRecentError
+	return
+}
+
+func (s *Session) BroadcastToSet(msg Message, uids StringSet) (lastErr error) {
+	s.Lock.Lock()
+	defer s.Lock.Unlock()
+	for uid := range uids {
+		if err := websocket.JSON.Send(s.Conns[uid].Conn, msg); err != nil {
+			lastErr = err
+		}
+	}
+	return
+}
+
+func (s *Session) RequestAccess(uid string, msg Message) error {
+	if msg.Author == s.Teacher {
+		if conn, ok := s.Conns[msg.Target]; ok {
+			conn.Subscriptions[uid] = true
+		} else {
+			return s.BroadcastError(uid, "Student does not exist")
+		}
+	} else {
+		return s.BroadcastError(uid, "Teacher permission required to request access to student")
+	}
+	return nil
 }
```

Comment:
Sorry, what exactly do you mean by this?

Reply:
I meant it in the parallel sense. Since this function mutates the actual structure, we should lock the data type. Something like:

```go
func (s *Session) RequestAccess(uid string, msg Message) error {
	s.Lock()
	defer s.Unlock()
	if msg.Author == s.Teacher {
		// ...
```

Will do the trick.